tests/testapp/tests/test_connection.py

```diff
@@ -0,0 +1,134 @@
+
+    @mock_patch_decorator
+    def test_get_remote_certs(self):
+        # mock certs being returned by server
+        certs = self.subset_cert.get_ancestors(include_self=True)
+        cert_serialized = json.dumps(CertificateSerializer(certs, many=True).data)
+        Connection._request.return_value = mock.Mock(data=cert_serialized)
+
+        # we want to see if the models are created (not saved) successfully
+        remote_certs = self.network_connection.get_remote_certificates('abc')
+        self.assertSetEqual(set(certs), set(remote_certs))
+
+    @mock_patch_decorator
+    def test_csr(self):
+        # mock a "signed" cert being returned by server
+        cert_serialized = json.dumps(CertificateSerializer(self.subset_cert).data)
+        Connection._request.return_value = mock.Mock(data=cert_serialized)
+        self.subset_cert.delete()
+
+        # we only want to make sure the "signed" cert is saved
+        with mock.patch.object(Key, "get_private_key_string", return_value=self.subset_cert.private_key.get_private_key_string()):
+            self.network_connection.certificate_signing_request(self.root_cert, '', '')
+        self.assertTrue(Certificate.objects.filter(id=json.loads(cert_serialized)['id']).exists())
+
+    @mock_patch_decorator
+    def test_get_cert_chain(self):
+        # mock a cert chain being returned by server
+        certs = self.subset_cert.get_ancestors(include_self=True)
+        cert_serialized = json.dumps(CertificateSerializer(certs, many=True).data)
+        Connection._request.return_value = mock.Mock(data=cert_serialized)
+        Certificate.objects.all().delete()
+
+        # we only want to make sure the cert chain is saved
+        self.network_connection._get_certificate_chain(certs[1])
+        self.assertEqual(Certificate.objects.count(), certs.count())
+
+    @mock_patch_decorator
+    def test_create_transfer_session(self):
+        session = SyncSession.objects.create(id=uuid.uuid4().hex, profile="profile", last_activity_timestamp=timezone.now())
+        self.network_connection.sync_session = session
+
+        self.assertEqual(TransferSession.objects.filter(active=True).count(), 0)
+        self.network_connection._create_transfer_session(True, ['ok'])
+        self.assertEqual(TransferSession.objects.filter(active=True).count(), 1)
+
+    @mock_patch_decorator
+    def test_close_transfer_session(self):
+        session = SyncSession.objects.create(id=uuid.uuid4().hex, profile="profile", last_activity_timestamp=timezone.now())
+        self.network_connection.sync_session = session
+        transfer = TransferSession.objects.create(id=uuid.uuid4().hex, filter="", push=True, sync_session=session, last_activity_timestamp=timezone.now())
+        self.network_connection.current_transfer_session = transfer
+
+        self.assertEqual(TransferSession.objects.filter(active=True).count(), 1)
+        self.network_connection._close_transfer_session()
```

Comment:
This is tricky, because it basically tries to delete it twice, once through the ORM and once through the API, right? So we're not fully testing both paths. We could perhaps assert that certain mocked methods are called once, or perhaps we leave fuller testing for "multi-server" mode tests.

Reply:
will just assert the mock method is called.